misc/vscode-extensions: improve scripts for updating VSC extensions

cideM opened the pull request:
This PR is a bit out of the blue but I've added my motivation for it under the appropriate heading. It was a fairly spontaneous thing that came up when I was migrating from controlling VSC extensions with VSC to Nix.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

The current `update_installed_exts.sh` script does several things. It tries to find the Visual Studio Code (VSC) binary (although it misses `-insiders`), and then it lists the extensions and prints a Nix object for each extensions which contains the updated version and sha. The idea is that it makes it easy to migrate to having extensions controlled with Nix.

Unfortunately this makes it hard to reuse the functionality in that script in other settings. What if I just want to update a single extension?

I thought it would be nice to just separate the two concerns. As such I've created two separate scripts which are used in the  `update_installed_exts.sh` script, so that it's 100% backwards compatible. The new `update_exts.sh` script makes it easier to e.g., pipe the name of an extension to it.

```
# Examples:
#
# $ echo 'ms-toolsai.jupyter' | ./update_exts.sh
#   {
#     name = "jupyter";
#     publisher = "ms-toolsai";
#     version = "2020.12.414227025";
#     sha256 = "1zv5p37qsmp2ycdaizb987b3jw45604vakasrggqk36wkhb4bn1v";
#   }
#
# $ ./list_installed_exts.sh | ./update_exts.sh
# ...lots of output as shown above
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cideM committed:
misc/vscode-extensions: improve scripts for updating extensions

The current update_installed_exts.sh always assumes that you're trying
update all installed extensions. That makes it hard to use the
get_vsixpkg function in a different setting.

This commit removes listing extensions, since it's trivial for a user to
run `code --list-extensions` but keeps and modifies the update script,
so it's more flexible.

Documentation was added to explain how the script works and explain VSC
in Nix(OS) in general.

The actual code is mostly the same.

SuperSandro2000 commented on pkgs/misc/vscode-extensions/update_exts.sh:
```diff
@@ -0,0 +1,65 @@
+  SHA=$(nix-hash --flat --base32 --type sha256 "$EXTTMP/$line.zip")
+
+  # Clean up.
+  rm -Rf "$EXTTMP"
+  # I don't like 'rm -Rf' lurking in my scripts but this seems appropriate.
```
```suggestion
  rm -rf "$EXTTMP"
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/update_exts.sh:
```diff
@@ -0,0 +1,65 @@
+function clean_up() {
+    TDIR="${TMPDIR:-/tmp}"
+    echo "Script killed, cleaning up tmpdirs: $TDIR/vscode_exts_*" >&2
+    rm -Rf "$TDIR/vscode_exts_*"
```
```suggestion
    rm -rf "$TDIR/vscode_exts_*"
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/list_installed_exts.sh:
```diff
@@ -0,0 +1,17 @@
+# See if we can find our `code` binary somewhere.
+# If the user passed in positional parameters, assume that the first is the
+# path to the code binary. Otherwise try 'code' and 'code-insiders'.
+if [ $# -ne 0 ]; then
```
```suggestion
if [[ $# -ne 0 ]]; then
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/list_installed_exts.sh:
```diff
@@ -0,0 +1,17 @@
+if [ $# -ne 0 ]; then
+    CODE=$1
+else
+    CODE=$(command -v code || command -v code-insiders)
```
please usw which

cideM replied:
What would you switch on though? The `$#` part? There's doesn't seem to be a single value I can switch on as far as I can see.

SuperSandro2000 replied:
command also returns exit code 0 when a function named code is defined which which does not.

cideM replied:
Oh wow sorry I somehow read "switch" instead of "which". So I just kept 90% of the code from the way it was before and there just used `command` without giving it too much thought to be honest

SuperSandro2000 replied:
which and command shouldn't matter here because my dotfiles shouldn't pollute the update script.

SuperSandro2000 commented on pkgs/misc/vscode-extensions/list_installed_exts.sh:
```diff
@@ -0,0 +1,17 @@
+    CODE=$(command -v code || command -v code-insiders)
+fi
+
+if [ -z "$CODE" ]; then
```
```suggestion
if [[ -z $CODE ]]; then
```

cideM commented:
I also need to make sure that the other shell scripts are called relative to the first file, not relative to the working directory

SuperSandro2000 commented:
> I also need to make sure that the other shell scripts are called relative to the first file, not relative to the working directory. Or maybe it's OK to use relative paths here as long as it's mentioned in the Wiki and the script itself

That should be documented in here https://github.com/NixOS/nixpkgs/tree/master/doc/languages-frameworks

cideM commented:
Haven't forgotten was just busy over the weekend, will do the necessary changes over the next couple of days

cideM commented:
I had a change of heart and decided to remove the existing script which lists installed extensions. After extracting the other bit it became apparent that listing extensions is a bit of bash when the user can just call `code --list-extensions`. I don't think abstracting that away helps.

I also added documentation, as was asked. I took some cues from the existing info in the Wiki:

- https://nixos.wiki/wiki/Vscode
- https://nixos.wiki/wiki/VSCodium

Please let me know if that's in the right direction 🙏

SuperSandro2000 commented on doc/languages-frameworks/vsc.md:
```diff
@@ -0,0 +1,191 @@
+
+### Extensions Helper
+
+The script at `https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vscode/vscode.nix#L28` makes updating marketplace extensions much more convenient. It takes a list of extensions on STDIN and outputs objects that you can paste into your Nix code.
```
Linking lines will likely break in the future.

cideM replied:
Hm yeah that's a good point I'll take that out

SuperSandro2000 reviewed (commented):
The examples are quite long. Maybe they could be reduced to the parts that matter for that section?

SuperSandro2000 commented:
Not sure why the manual fails to build.

cideM commented:
@SuperSandro2000 

I updated the PR. I removed all line links. I also overhauled the documentation so it has less superfluous code snippets and gets to the point a bit quicker.

SuperSandro2000 commented:
> Hopefully fixed doc build by giving the file a valid name

It did.

SuperSandro2000 commented on doc/languages-frameworks/vscode.section.md:
```diff
@@ -0,0 +1,121 @@
+
+Lastly, you can also install extensions imperatively (through the VSC UI) just like you would without Nix. This works best if you're not using NixOS, since, on NixOS, it'll fail for extensions such as `vsliveshare`.
+
+The resulting list of extensions can be passed to both `vscodeExtensions` and the `extensions` option from HM. It's recommended to use `vscodeExtensions` though, since the HM option uses an older method of handling extensions, which is not as reliable (see this [issue](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&query=vscode-extensions)).
```
You did not link the issue but the search page for vscode extensions.

cideM replied:
:facepalm:

SuperSandro2000 commented:
Please also add an entry to https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/index.xml and yourself as a codeowner for ``doc/languages-frameworks/vscode.section.md``.

cideM commented:
- Added myself to CODEOWNERS for vscode docs
- Fixed link
- Added index entry

cideM commented:
Accidentally removed the branch from my fork

SuperSandro2000 commented:
@jonringer can you take a look at this?

jonringer commented on doc/languages-frameworks/vscode.section.md:
```diff
@@ -0,0 +1,121 @@
+    version = "0.47.2";
+    sha256 = "1hp6gjh4xp2m1xlm1jsdzxw9d8frkiidhph6nvl24d0h8z34w49g";
+  }];
+  vscode-with-extensions = pkgs.vscode-with-extensions.override {
+    vscodeExtensions = extensions;
+  };
+in
+{
+  environment.systemPackages = [
+    vscode-with-extensions
```
I would prefer not to use `vscode-with-extensions` twice
```suggestion
  vscode-with-my-extensions = pkgs.vscode-with-extensions.override {
    vscodeExtensions = extensions;
  };
in
{
  environment.systemPackages = [
    vscode-with-my-extensions
```

cideM replied:
I'll take care of this later today and also add some info about your recent PRs about buildFSHUserEnv for VSCode, thanks fo the review :confetti_ball:

jonringer reviewed (commented):
otherwise looks mostly good to moe

ShamrockLee commented:
@cideM @jonringer I'll integrate #121583 with this commit after this PR is merged.

Please take a look at that PR then.

cideM commented:
@jonringer I renamed the attribute that was sort of shadowing the package and I also added a paragraph about your FHS changes. It's somewhat duplicated from the Wiki but I'm in general unsure what goes where with regards to Wiki and the official documentation.

i also rearranged some headings to make the different modes of operation a bit more clear. You can ultimately use Nix or the VSC UI and I think that's gonna be the most important question people have, so it should be easy to find.

cideM commented:
ping @jonringer  (or should I ping someone else?)

jonringer commented:
I think this now conflicts with  https://github.com/NixOS/nixpkgs/pull/127741.

We should probably have merged this long ago, sorry

cideM commented:
Ah no worries, I'll just close this one then and see if the documentation changes can help with the other PR at least
